### PR TITLE
Project Start - Code review

### PR DESCRIPTION
**Pull request by eodbrent**
code review

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+
+def ace_value(hand, hand_value):
+    """Determines Ace = 1 or 11. If 11 pushes hand over 21, Ace value = 1"""
+    hand_ace_value = hand_value
+    for card in hand:
+        if card[0] == 'A':
+            #  For each Ace, subtract 10
+            hand_ace_value -= 10
+    if hand_value > 21 >= hand_ace_value:
```
Comment (Choochoo) [inline main.py]: Trusting you know this works, tried to think about this logic and I got dizzy.

File: main.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        user_value = get_hand_value(user_hand)
+        print(f"{PURPLE}Your hand is worth: {user_value}{RESET}")
+        print(f"{DEALER}{YELLOW} hand: {dealer_hand}{RESET}")
+        dealer_value = get_hand_value(dealer_hand)
+        print(f"{DEALER}{YELLOW} hand is worth: {dealer_value}{RESET}")
+        print("What would you like to do?")
+        game_on = True
+        while game_on:
+            user_turn = True
+            while user_turn:
```
Comment (Choochoo) [inline main.py]: Move the second while to something like game_loop, then another loop in there.

File: main.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                if user_value == 21:
```
Comment (Choochoo) [inline main.py]: Move all 21's to a const then refer to it for easier readability

GAME_WINNING_SCORE = 21, something to that effect so it can read like a book "if SCORE == GAME_WINNING_SCORE:"

File: main.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                    user_turn = False
+                    break
+                elif user_value > 21:
+                    print("You busted!")
+                    user_turn = False
+                    break
+                display_running_count()
+                choice = input("Hit (h), Stand (s), Double Down (d), Split (Requires doubles) (t), Quit (q), \"deck\" displays number of cards left: ").lower()
+
+                if choice == "q":
```
Comment (Choochoo) [inline main.py:174]: these are called magic strings, all magic strings should be set to a constant for easier readability, why is choice going to be "q" here, I see "Q" in the above string, but maybe that above string should also be a const.
it is sorta like controller mapping "Quit (q)" should be "Quit ({QUIT_KEY})"

File: main.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                elif choice == "h":
+                    print("hit")
+                    if len(shuffled_deck) > 0:
+                        user_hand.append(shuffled_deck[0])
+                        shuffled_deck.remove(shuffled_deck[0])
+                        user_value = get_hand_value(user_hand)
+                        for card in user_hand:
+                            if card[0] == 'A':
+                                user_value = ace_value(user_hand, user_value)
+                                print("You had an Ace. We just checked if it needed to be 11 or 1")
```
Comment (Choochoo) [inline main.py]: I don't think these one off print lines occurrences need to be in a const, but if you wanted to do multi language you could set them all part of a json file (or whatever type you choose, XML, YAML), then just convert that one file for all languages, on load it sets all these lines to whatever language, that'd be very cool feature and easy conversion using AI, it could even output all the files for all different languages once you get english setup and loading.  Just an idea.

File: main.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                        print(f"{YELLOW}Dealer's{RESET} hand: {dealer_hand} at {dealer_value}")
+                    elif dealer_value > 21:
+                        print(f"{DEALER} busts. {PLAYER} wins!")
+                        dealer_turn = False
+                    elif user_value < dealer_value <= 21:
+                        print(f"{DEALER} wins.")
+                        dealer_turn = False
+                        break
+            game_on = False
+            a = "another"
```
Comment (Choochoo) [inline main.py]: Why are these here? if they need to be here it may need a comment, a = "another" is confusing.

File: main.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                    break
+                elif choice == 'd':
+                    print("Double Down - not functional yet")
+                elif choice == 't':
+                    print("Split - not functional yet")
+                else:
+                    print("{RED}Invalid choice. Please try again.")
+            dealer_turn = True
+            if user_turn:
+                while dealer_turn:
```
Comment (Choochoo) [inline main.py:201]: make this into a function.
def dealer_turn:

inside that create this while loop, then instead of all these if's should be a return i.e.:

```
dealer_needs_to_draw = dealer_value < 21 and dealer_value <= user_value
if dealer_needs_to_draw:
      dealer_draw()
      return true
dealer_is_over = dealer_value > 21
if dealer_is_over
   print
   return false
```
   
//LAST "IF" NOT NEEDED, it should meet all cases and return the "default" value.

File: main.py (hunk 4 above)
Comment (Choochoo) [inline main.py:174]: Also these if/elses need to go into a main function and any inner logic needs to have its own function, a function needs to do 1 thing i.e.
```
def body
    is_left_arm_ok()
    is_right_arm_ok()
 
 def is_left_arm_ok()
       is_elbow_ok()
       is_hand_ok()
```

Comment (Choochoo) [reply main.py:174]: Also once it meets an "if" case, return immediately. I have an example below of that.

**Review by Choochoo — COMMENTED** [review]
Nice logic, logic seems perfect just some refactoring of ideas to be better readable.

File: main.py (hunk 2 above)
Comment (eodbrent) [reply main.py]: I agree, improvement is needed on these while loops. However, please elaborate on your suggestion, line 103 has 'while main_game_loop:' to handle initial user decisions and game setup.

File: main.py (hunk 3 above)
Comment (eodbrent) [reply main.py]: I will implement this change. Good catch, thank you!

File: main.py (hunk 4 above)
Comment (eodbrent) [reply main.py:174]: I'll look into magic strings and play around with this. I like this idea!

File: main.py (hunk 5 above)
Comment (eodbrent) [reply main.py]: This does sound neat. I'll talk more about it offline with you, and then probably add it as a feature/enhancement when I'm ready to progress to adding JSON etc. This particular print statement probably should have been taken out or reworded.

File: main.py (hunk 6 above)
Comment (eodbrent) [reply main.py]: Good catch, I'll add comments for readability. It's for rewording the introduction question - How about "a" game of blackjack? Or, a = "another" - How about "another" game of blackjack.

File: main.py (hunk 1 above)
Comment (eodbrent) [reply main.py]: This was originally, if hand_value > 21 and hand_ace_value <= 21. This handles whether or not an Ace should be counted as 1 point or 11. Pycharm suggested shortening the logic to what you see. I agree, it's dizzying. I'll likely update this logic to give the user more of a choice when their Aces are worth 1 or 11 and do something about this readability.

File: main.py (hunk 7 above)
Comment (eodbrent) [reply main.py:201]: I like this idea, I think it will help when I'm ready to shift everything to an entry point function.

File: main.py (hunk 4 above)
Comment (eodbrent) [reply main.py:174]: I might have responded to this one in your previous comment. The intent is to eventually move to a main program function.

File: main.py (hunk 6 above)
Comment (Choochoo) [reply main.py]: you remove this variable completely and set this to 
first_game_finished = true
"Would you like to play a" + (first_game_finished ? "nother " : "") + "game?

File: main.py (hunk 5 above)
Comment (eodbrent) [reply main.py]: Cleaned this up a lot! Fixed the Ace logic, made it cleaner, got rid of a lot of unnecessary checks in the main loop and relevant functions. Now, hand_with_ace_value replaced ace_value and is called from get_hand_value with improved logic.

File: main.py (hunk 7 above)
Comment (eodbrent) [reply main.py:201]: I was reviewing your reviews, and I just noticed your use of booleans: dealer_is_over = dealer_value > 21. I was unaware of this Python capability. Very cool!

Comment (Choochoo) [reply main.py:201]: yes because it reads more like a story you're creating.
"dealer_value < 21 and dealer_value <= user_value"  makes another coder (or yourself in the future) have to think about what its doing, if you set it to a bool, then that is self-documenting code, no need for comments, and it reads like a conversation
```
dealer_needs_to_draw = dealer_value < 21 and dealer_value <= user_value
if dealer_needs_to_draw:
```
You should only comment your code as a last resort, function and variable names should be able to do 90%
